### PR TITLE
Add text component

### PR DESCRIPTION
**Pull request by EternityForest**
# What does this implement/fix?
This is an updated version of the text inputs component branch, building on work by [mauritskorse](https://github.com/mauritskorse) in https://github.com/esphome/esphome/pull/3638 which seems to have not gotten any updates for a few months.

Almost all of the work here is his, and he deserves full credit, I have only solved some merge conflicts and made a few minor changes as mentioned in the review, as I had some free time and he seems to be busy lately.

The original description is here:

> Adding a input text component (similar to [Home Assistant](https://www.home-assistant.io/integrations/input_text/)). It allows for setting a string by the user through webserver comp front-end (v1). Implementation for v2 webserver front-end and Home Assistant is not done yet.

But I have since added V2 support(https://github.com/esphome/esphome-webserver/pull/43), and I believe that I have covered most of the requested changes in the original.

Changes relative to the original:

Synced to upstream

- Changed the message type numbers as the old ones conflicted with newer upstream work
- Added the ability to persist strings to flash
- Use this-> for property references instead of the implicit ones
- add validate_ method(I wasn't quite sure what was actually needed or how validate works, more of a draft there, maybe @jesserockz can weigh in on if I did it right?)
- Deleted commented out code that did not appear to be relevant
- value_ begins at nullpointer by default(Not sure if that's actually a thing, or if the other components do it for a reason?)
- Added min_length, max_length, and pattern to the JSON in the web server.

The protocol buffers here were **not** compiled with 3.6 as mentioned in a comment, because that version is not available in newer Ubuntus.  If it's an exact version requirement and not a "Any later minor version" kind of thing, I can set up a VM and fix this.

Replaces and closes #3638

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3293

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
  - platform: template
    name: "test 2 text"
    id: "it_2"
    icon: "mdi:text"
    optimistic: true
    internal: false
    initial_value: "Welcome ESPHOME"
    restore_value: true
    max_restore_data_length: 24

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

(The original PR is marked as having documentation, but I have not updated it to include the newly-fixed restore_value)

### CODE REVIEW

File: esphome/components/api/api.proto [hunk 1]
```diff
@@ -1517,3 +1518,49 @@ message AlarmControlPanelCommandRequest {
   AlarmControlPanelStateCommand command = 2;
   string code = 3;
 }
+
+// ===================== TEXT =====================
+enum TextMode {
+  TEXT_MODE_AUTO = 0;
```
Comment (jesserockz) [inline esphome/components/api/api.proto]: I don't see how an auto mode will help at all..
HA doesn't seem to have this as a mode either

File: esphome/components/text/__init__.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+# Conditions
+TextMode = text_ns.enum("TextMode")
+
+TEXT_MODES = {
+    "AUTO": TextMode.TEXT_MODE_AUTO,
+    "TEXT": TextMode.TEXT_MODE_TEXT,
+    "PASSWORD": TextMode.TEXT_MODE_PASSWORD,  # to be implemented for keys, passwords, etc.
+}
+
+icon = cv.icon
```
Comment (jesserockz) [inline esphome/components/text/__init__.py]:
```suggestion
```

File: esphome/components/text/__init__.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+# Actions
+TextSetAction = text_ns.class_("TextSetAction", automation.Action)
+
+# Conditions
+TextMode = text_ns.enum("TextMode")
+
+TEXT_MODES = {
+    "AUTO": TextMode.TEXT_MODE_AUTO,
```
Comment (jesserockz) [inline esphome/components/text/__init__.py]:
```suggestion
```

File: esphome/components/text/__init__.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+TEXT_SCHEMA = cv.ENTITY_BASE_SCHEMA.extend(cv.MQTT_COMPONENT_SCHEMA).extend(
+    {
+        cv.OnlyWith(CONF_MQTT_ID, "mqtt"): cv.declare_id(mqtt.MQTTTextComponent),
+        cv.GenerateID(): cv.declare_id(Text),
+        cv.Optional(CONF_ON_VALUE): automation.validate_automation(
+            {
+                cv.GenerateID(CONF_TRIGGER_ID): cv.declare_id(TextStateTrigger),
+            }
+        ),
+        cv.Optional(CONF_MODE, default="AUTO"): cv.enum(TEXT_MODES, upper=True),
```
Comment (jesserockz) [inline esphome/components/text/__init__.py]:
```suggestion
        cv.Required(CONF_MODE): cv.enum(TEXT_MODES, upper=True),
```

File: esphome/components/text/text_call.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  if (sz < this->parent_->traits.get_min_length()) {
+    ESP_LOGW(TAG, "'%s' - Value set for TextCall is too short", name);
+    this->value_.reset();
+    return;
+  }
+}
+
+void TextCall::perform() {
```
Comment (jesserockz) [inline esphome/components/text/text_call.cpp:38]: Missing the call to validate_ ?

File: esphome/components/text/text_traits.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include "esphome/core/log.h"
```
Comment (jesserockz) [inline esphome/components/text/text_traits.cpp]: Not much point in this file...

File: esphome/components/text/text_traits.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "esphome/core/helpers.h"
+
+namespace esphome {
+namespace text {
+
+enum TextMode : uint8_t {
+  TEXT_MODE_AUTO = 0,
+  TEXT_MODE_TEXT = 1,
+  TEXT_MODE_PASSWORD = 2,
```
Comment (jesserockz) [inline esphome/components/text/text_traits.h]:
```suggestion
  TEXT_MODE_TEXT = 0,
  TEXT_MODE_PASSWORD = 1,
```

File: tests/component_tests/text/test_text.yaml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+esphome:
```
Comment (jesserockz) [inline tests/component_tests/text/test_text.yaml:1]: These tests are not used by CI. 
Please add to one of the test yaml files in the tests folder.

**Review by jesserockz — CHANGES_REQUESTED** [review]
Overall looks fine. I have not tested yet, but will hopefully this week.
A few things below.

I don't think we need to have an auto mode (there is no way to determine if you want text or password), and mode should be required not optional.

**Comment by EternityForest** [comment]
Yeah, I'm not exactly sure what the intent of the auto mode was, but I can
certainly get back on this PR and make those changes!

I could see the value of making generic text as the default and having mode
be optional, but then again, explicit is better than implicit.

I'm impressed by the thoroughness of the review and attention to code
quality!

**Comment by jesserockz** [comment]
> I'm impressed by the thoroughness of the review and attention to code
quality!

When given the time to sit there and without distractions, its really nice to be able to sit there and review code from top to bottom.

Hopefully we can get this into the next ESPHome release, I might be able to get it into the HA November release if I work on that side tomorrow as the beta starts tomorrow for that.

**Comment by nielsnl68** [comment]
I will give you a peace of 🍰 when this will be finally in 👯‍♂️

**Review by jesserockz — APPROVED** [review]

**Pull request merged**